### PR TITLE
MVI FindProfileMessage log missing keys on raise

### PR DESCRIPTION
**Pull request by f1337**
# Wat

To help Identity, et al, trace SSOe interaction errors.

## Before

We would raise an error with `required keys are missing` if any required profile keys were missing, and `required values are missing` if any required profile keys were present but had empty values. We did not however indicate which keys were missing, or were missing values.

## After

We now indicate which keys were missing, or were missing values, in the error message.

### CODE REVIEW

File: spec/lib/mpi/messages/find_profile_message_fields_spec.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  describe '.valid?' do
+    subject { described_class.new(profile) }
+
+    let(:missing_keys) { %i[given_names last_name birth_date ssn] }
+
+    before do
+      subject.validate
+    end
+
+    context 'missing keys and values' do
```
Comment (kfrz) [inline spec/lib/mpi/messages/find_profile_message_fields_spec.rb:16]:
```suggestion
    context 'with missing keys and values' do
```

File: spec/lib/mpi/messages/find_profile_message_fields_spec.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      let(:profile) { {} }
+
+      its(:valid?) { is_expected.to be(false) }
+      its(:missing_keys) { is_expected.to be(true) }
+      its(:missing_values) { is_expected.to be(true) }
+    end
+
+    context 'missing values' do
```
Comment (kfrz) [inline spec/lib/mpi/messages/find_profile_message_fields_spec.rb:24]:
```suggestion
    context 'with missing values' do
```

File: lib/mpi/messages/find_profile_message.rb [hunk 3]
```diff
@@ -38,8 +39,13 @@ def initialize(profile, orch_search = false, edipi = nil)
...
       def required_fields_present?(profile)
-        raise ArgumentError, 'required keys are missing' unless REQUIRED_FIELDS.all? { |k| profile.key?(k) }
-        raise ArgumentError, 'required values are missing' unless REQUIRED_FIELDS.all? { |k| profile[k].present? }
+        fields = FindProfileMessageFields.new(profile)
+        fields.validate
+
+        raise ArgumentError, "required keys are missing: #{fields.missing_keys}" if fields.missing_keys.present?
+        if fields.missing_values.present?
+          raise ArgumentError, "required values are missing for keys: #{fields.missing_values}"
+        end
```
Comment (f1337) [inline lib/mpi/messages/find_profile_message.rb:48]: One is a 1-liner, the other 3, "because Rubocop".

File: spec/lib/mpi/messages/find_profile_message_fields_spec.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      let(:profile) { { given_names: nil, last_name: '', birth_date: nil, ssn: '' } }
+
+      its(:valid?) { is_expected.to be(false) }
+      its(:missing_keys) { is_expected.to be(false) }
+      its(:missing_values) { is_expected.to be(true) }
+    end
+
+    context 'valid-ish' do
```
Comment (kfrz) [inline spec/lib/mpi/messages/find_profile_message_fields_spec.rb:32]:
```suggestion
    context 'with valid-ish attributes' do
```

Not sure about what's `valid-ish` here

File: Gemfile [hunk 5]
```diff
@@ -170,6 +170,7 @@ group :development, :test do
   gem 'rack-vcr'
   gem 'rainbow' # Used to colorize output for rake tasks
   gem 'rspec-instrumentation-matcher'
+  gem 'rspec-its'
```
Comment (f1337) [inline Gemfile:173]: I _could_ split this addition into its own PR, but that feels like busy work TBH.

**Review by kfrz — CHANGES_REQUESTED** [review]
Minor changes for the spec

**Comment by f1337** [comment]
@kfrz I updated the messages, could you please unblock? Thank you!

File: spec/lib/mpi/messages/find_profile_message_spec.rb [hunk 6]
```diff
@@ -149,7 +149,29 @@
...
+      end
+    end
+  end
+
+  describe '#required_fields_present?' do
+    subject { described_class.new(profile) }
+
+    let(:missing_keys) { ':given_names, :last_name, :birth_date, :ssn' }
+
+    context 'missing keys' do
```
Comment (kfrz) [inline spec/lib/mpi/messages/find_profile_message_spec.rb:162]:
```suggestion
    context 'with missing keys' do
```

File: spec/lib/mpi/messages/find_profile_message_spec.rb [hunk 7]
```diff
@@ -149,7 +149,29 @@
...
+      let(:profile) { {} }
+
+      it 'raises with list of missing keys' do
+        expect { subject }.to raise_error(/#{missing_keys}/)
+      end
+    end
+
+    context 'missing values' do
```
Comment (kfrz) [inline spec/lib/mpi/messages/find_profile_message_spec.rb:170]:
```suggestion
    context 'with missing values' do
```

**Review by kfrz — APPROVED** [review]

**Review by bramleyjl — COMMENTED** [review]
looks good to me
